OpenRA.Mods.Common/Traits/Infantry/TerrainModifiesDamage.cs

```diff
@@ -0,0 +1,71 @@
+﻿ #region Copyright & License Information
+ /*
+  * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+  * This file is part of OpenRA, which is free software. It is made
+  * available to you under the terms of the GNU General Public License
+  * as published by the Free Software Foundation. For more information,
+  * see COPYING.
+  */
+ #endregion
+
+using System;
+using System.Linq;
+using OpenRA.GameRules;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common
+{
+	public class TerrainModifiesDamageInfo : ITraitInfo
+	{
+		[Desc("An array of terrain types that modify this actor's incoming damage, if the actor is standing on these types.")]
+		public readonly string[] TerrainTypes = { };
+
+		[Desc("Total damage percent for each TerrainType. 120 = 120%, 80 = 80%, 50 = 50%, etc.")]
+		public readonly int[] DamagePerType = { };
+
+		[Desc("Modify healing damage? For example: A friendly medic.")]
+		public readonly bool ModifyHealing = false;
+
+		public object Create(ActorInitializer init) { return new TerrainModifiesDamage(init.Self, this); }
+	}
+
+	public class TerrainModifiesDamage : IDamageModifier
+	{
+		public readonly Actor Self;
+		public readonly TerrainModifiesDamageInfo Info;
+
+		public TerrainModifiesDamage(Actor self, TerrainModifiesDamageInfo info)
+		{
+			Self = self;
+			Info = info;
+
+			if (info.TerrainTypes.Length != info.DamagePerType.Length)
+				throw new YamlException("TerrainModifiesDamage's properties do not have the same number of items for {0}.".F(self.Info.Name));
+		}
+
+		public int GetDamageModifier(Actor attacker, DamageWarhead warhead)
+		{
+			var damage = 100;
+			if (attacker.Owner.IsAlliedWith(Self.Owner) && warhead.Damage < 0 && !Info.ModifyHealing)
+				return damage;
+
+			var world = Self.World;
+			var map = world.Map;
+			var tileSet = world.TileSet;
+
+			var tiles = map.MapTiles.Value;
+			var pos = map.CellContaining(Self.CenterPosition);
+			var tileIndex = tileSet.GetTerrainIndex(tiles[pos]);
+			var terrainType = tileSet[tileIndex].Type;
```

Comment:
I try to limit the expression count in an indexer and param list as much as possible, but you have a point that this isn't so bad. I'll clean it up a little.

I plan on exposing this data (TileSetInfo of the actor's current cell) through an Actor extension method so this may be moved around a bit in the end.